### PR TITLE
[FLINK-4692] [tableApi] Add tumbling group-windows for batch tables

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetWindowAggregateRule.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.calcite.rel.convert.ConverterRule
+import org.apache.flink.api.table.TableException
+import org.apache.flink.api.table.plan.logical.rel.LogicalWindowAggregate
+import org.apache.flink.api.table.plan.nodes.dataset.{DataSetConvention, DataSetWindowAggregate}
+
+import scala.collection.JavaConversions._
+
+class DataSetWindowAggregateRule
+  extends ConverterRule(
+      classOf[LogicalWindowAggregate],
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetWindowAggregateRule.scala:31]: 缩进

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala [hunk 2]
```diff
@@ -84,7 +84,9 @@ case class ProcessingTimeTumblingGroupWindow(
       resolve(size))
 
   override def validate(tableEnv: TableEnvironment): ValidationResult =
-    super.validate(tableEnv).orElse(TumblingGroupWindow.validate(tableEnv, size))
+    super.validate(tableEnv)
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala]: 建议不再这里做。到抽象类里面做。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala [hunk 3]
```diff
@@ -173,7 +176,9 @@ case class ProcessingTimeSlidingGroupWindow(
       resolve(slide))
 
   override def validate(tableEnv: TableEnvironment): ValidationResult =
-    super.validate(tableEnv).orElse(SlidingGroupWindow.validate(tableEnv, size, slide))
+    super.validate(tableEnv)
+      .orElse(SlidingGroupWindow.validate(tableEnv, size, slide))
+      .orElse(ProcessingTimeWindow.validate(tableEnv, size))
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala]: 同上

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala [hunk 4]
```diff
@@ -196,7 +201,8 @@ case class EventTimeSlidingGroupWindow(
     super.validate(tableEnv)
       .orElse(SlidingGroupWindow.validate(tableEnv, size, slide))
       .orElse(size match {
-        case Literal(_, RowIntervalTypeInfo.INTERVAL_ROWS) =>
+        case Literal(_, RowIntervalTypeInfo.INTERVAL_ROWS)
+          if tableEnv.isInstanceOf[StreamTableEnvironment] =>
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:215]: 同上

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala [hunk 5]
```diff
@@ -232,7 +238,9 @@ case class ProcessingTimeSessionGroupWindow(
       resolve(gap))
 
   override def validate(tableEnv: TableEnvironment): ValidationResult =
-    super.validate(tableEnv).orElse(SessionGroupWindow.validate(tableEnv, gap))
+    super.validate(tableEnv)
+      .orElse(SessionGroupWindow.validate(tableEnv, gap))
+      .orElse(ProcessingTimeWindow.validate(tableEnv, gap))
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala]: 同上

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala [hunk 6]
```diff
@@ -256,3 +264,14 @@ case class EventTimeSessionGroupWindow(
 
   override def toString: String = s"EventTimeSessionGroupWindow($name, $timeField, $gap)"
 }
+
+object ProcessingTimeWindow {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala]: 若可以在抽象类做，这里就不需要了。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    }
+  }
+
+  /**
+    * get window information
+    * @param groupWindow window
+    * @return window size (with unit),
+    *         and time field position & type information if it is an event-time window
+    */
+  private def getWindowInformation(groupWindow: LogicalWindow)
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala]: 需要明天讨论

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 8]
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
...
+    */
+  def createWindowPrepareMapFunction(
+    window: WindowInformation,
+    namedAggregates: Seq[CalcitePair[AggregateCall, String]],
+    groupings: Array[Int],
+    inputType: RelDataType): MapFunction[Any, Row] = {
+
+    // the addition one field is used to store time attribute
+    val (aggFieldIndexes, aggregates) = transformToAggregateFunctions(
+      namedAggregates.map(_.getKey), inputType, groupings.length + 1)
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala]: 不知道加明确的名字好?还是这样好

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/WindowAggregateMapFunction.scala [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+class WindowAggregateMapFunction(
+    private val aggregates: Array[Aggregate[_]],
+    private val aggFields: Array[Int],
+    private val groupingKeys: Array[Int],
+    private val timeFieldPos: Int,
+    private val timeFieldTypeInfo: TypeInformation[_],
+    private val windowSize: Long,
+    private val isTimeWindow: Boolean,
+    @transient private val returnType: TypeInformation[Row])
+    extends RichMapFunction[Row, Row] with ResultTypeQueryable[Row] {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/WindowAggregateMapFunction.scala]: 缩进

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 10]
```diff
@@ -649,9 +649,6 @@ class Table(
     * @return A windowed table.
     */
   def window(groupWindow: GroupWindow): GroupWindowedTable = {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:651]: 我想在这里检查是不是比较好！（该检查的都提前检查）

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+      .item("window", window)
+      .item(
+        "select", aggregationToString(
+          inputType,
+          grouping,
+          getRowType,
+          namedAggregates,
+          namedProperties))
+  }
+
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala:101]: 是不是应该有计算cost.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 12]
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
...
+    *
+    * NOTE: this function is only used for time based window on batch tables.
+    */
+  def createWindowPrepareMapFunction(
+    window: WindowInformation,
+    namedAggregates: Seq[CalcitePair[AggregateCall, String]],
+    groupings: Array[Int],
+    inputType: RelDataType): MapFunction[Any, Row] = {
+
+    // the addition one field is used to store time attribute
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala]: 是否可以根据window类型创建不同的mapfunction

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 13]
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
...
+
+    if (groupingOffsetMapping.length != groupings.length ||
+      aggOffsetMapping.length != namedAggregates.length) {
+      throw new TableException(
+        "Could not find output field in input data type " +
+          "or aggregate functions.")
+    }
+
+    if (window.isTimeWindow) {
+      // time window
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala]: 是否可以根据window类型创建不同的reduceGroupFunction.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 14]
```diff
@@ -308,7 +434,7 @@ object AggregateUtil {
   /**
     * Return true if all aggregates can be partially computed. False otherwise.
     */
-  private[flink] def doAllSupportPartialAggregation(
+  def doAllSupportPartialAggregation(
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:452]: 为啥去掉private[flink]

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 15]
```diff
@@ -84,14 +86,138 @@ object AggregateUtil {
...
+
   /**
     * Create a [[org.apache.flink.api.common.functions.GroupReduceFunction]] to compute aggregates.
     * If all aggregates support partial aggregation, the
     * [[org.apache.flink.api.common.functions.GroupReduceFunction]] implements
     * [[org.apache.flink.api.common.functions.CombineFunction]] as well.
     *
     */
-  private[flink] def createAggregateGroupReduceFunction(
+  def createAggregateGroupReduceFunction(
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:235]: 建议加上private[flink]

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 16]
```diff
@@ -722,9 +719,6 @@ class GroupedTable(
...
-    if (table.tableEnv.isInstanceOf[BatchTableEnvironment]) {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala:725]: 这里检查所有batch上支持的窗口类型是否更好些。该检查的都提前检查

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala (hunk 14 above)
Comment (wuchong) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:452]: `doAllSupportPartialAggregation` 以及 `createAggregateGroupReduceFunction`  都是该util提供给其他类使用的帮助函数，所以应该设置为public的可见性。`private[flink]` 应该用在该类是提供给用户的，而该方法不希望对用户可见的。但是 `AggregateUtil.scala` 是非面向用户的，所以不需要`private[flink]`。 如 `CodeGenerator` ， `CodeGenUtils` ， `UserDefinedFunctionUtils` 等等都是类似的。

**Pull request closed without merging**